rskj-core/src/main/java/org/ethereum/datasource/TransientMap.java

```diff
@@ -0,0 +1,212 @@
+        private transient Map<K, V> added;
+        private transient Map<K, V> changed;
+        private transient Set<K> deleted;
+
+        TransientMap(Map<? extends K, ? extends V> m) {
+            if (m == null) {
+                throw new NullPointerException();
+            } else {
+                this.m = (Map<K, V>) m;
+                added = new HashMap<>();
+                changed = new HashMap<>();
+                deleted = new HashSet<>();
+
+            }
+        }
+
+        public int size() {
+
+            if (m==null) {
+                return added.size();
+            } else {
+                // todo: mix
+                return this.m.size()-deleted.size()+added.size();
+            }
+        }
+
+        public boolean isEmpty() {
+            if (m==null)
+                return added.isEmpty();
+            return this.m.isEmpty() && added.isEmpty();
+        }
+
+        public boolean containsKey(Object key) {
+            if (added.containsKey(key))
+                return true;
+            if (m==null)
+                return false;
+            if (changed.containsKey(key))
+                return true;
+            return this.m.containsKey(key) && (!deleted.contains(key));
+
+        }
+
+        public boolean containsValue(Object val) {
+            throw new UnsupportedOperationException();
+        }
+
+        public V get(Object key) {
+            if (m==null)
+                return added.get(key);
+
+            if (deleted.contains(key))
+                return null;
+            V ret = added.get(key);
+            if (ret==null)
+                ret = changed.get(key);
+            if (ret==null)
+                ret = this.m.get(key);
+            return ret;
+        }
+
+        public V put(K key, V value) {
+            if (m==null)
+                return added.put(key,value);
+
+            V ret = m.get(key);
+            if (ret!=null) {
+                if (deleted.contains(key)) {
+                    deleted.remove(key);
+                    changed.put(key,value);
+                    return null;
+                }
+                return changed.put(key,value);
+            } else
+                return added.put(key,value);
+        }
+
+        public V remove(Object key) {
+            V ret;
+            if (m!=null) {
+                if (!deleted.contains(key)) {
+                    deleted.add((K) key);
+                    ret = changed.remove(key);
+                    if (ret==null)
+                        ret = added.remove(key);
+                    if (ret==null)
+                        ret = m.get(key);
+                } else
+                    // double delete
+                    ret = null;
+            } else {
+                ret = added.remove(key);
+            }
+            return ret;
+        }
+
+        public void putAll(Map<? extends K, ? extends V> am) {
+
+            if (m==null) {
+                return;
```

Comment:
Why are we exiting here if `m==null` and not doing the same for `put()`?

Reply:
And why not calling put() in the loop? Do we want a different behavior here?

Reply:
I've reused put() in putAll(), please open a new thread if you think we shouldn't @SergioDemianLerner.